### DISCUSSION TITLE
Drivers should stand on their own

### DISCUSSION THREAD

**Discussion by fabiques** [qilimanjaro-tech/qililab#272] [Ideas]
I would like to share an opinion, for us to discuss.

Every driver we write should stand on its own, independently of qililab

Connecting to a device should be as simple as:

```python
vna = E5080B('vna' , 192.168.1.230)
``` 

All functionallity should be available directly from it. 

```python
vna.set_power(power)
vna.set_center_frequency(f0)
``` 

In a sense all drivers we write should be publishable in QCODES, and I argue we should actually publish them.

Let me know what you think.

Cheers,

Fabio Henriques

**Comment by AlbertMitjans** [comment]
I agree! And all drivers currently supported in `qililab` should stand on its own! Let me know if this is not the case.

**Comment by fabiques** [comment]
The implementation of the VNA driver that Isaac wrote can be called like this:

```python
# Setup settings, you should only change 'power'
# (When doing initial connection you are not sending the command to the VNA to set the power to the specified so set it manually)
vna_settings = {"power": -20.0, "firmware": "A.15.10.06", "id_": 0, "category": "vna"}
vna_e5080b = VNA(vna_settings)

# Create an instance of the driver
vna_device = VNADriver("vna", address="192.168.1.254")
vna_e5080b.device = vna_device
``` 

Rather than:

```python
# Create an instance of the driver
vna_device = VNADriver("vna", address="192.168.1.254")
```

**Comment by AlbertMitjans** [comment]
I agree 100%!! We will work on this

**Comment by fabiques** [comment]
I would like to add a reasoning on this. Sometimes there are things that are implemented in the driver that would not be used in qililab, for example the ability of the yokogawa to measure IV curves. Qililab should never use something like that, but we might use it in other instances. And I argue we should no write two drivers.

Also, the same device can sometimes be used in diferent ways within qililab. The yokogawa can function both as a current and voltage source. Both are useful within qililab. Both should call the same driver.

**Reply by AlbertMitjans** [reply]
Thanks for the reasoning!! It helps a lot 🙌
